File: solution/174/solution.cpp
#include <stdio.h>
#include <vector>

class Solution {
public:
    int calculateMinimumHP(std::vector<std::vector<int> >& dungeon) {
        if(dungeon.size()==0)return 0;
        int row = dungeon.size();
        int column = dungeon[0].size();
        std::vector<std::vector<int> > dp(row, std::vector<int>(column, 0));
        dp[row-1][column-1] = std::max(1, 1 - dungeon[row-1][column-1]);
        for(int i = column-2;i>=0;i--){
            dp[row-1][i] = std::max(1, dp[row-1][i+1] - dungeon[row-1][i]);
        }
        for(int i = row-2;i>=0;i--){
            dp[i][column-1] = std::max(1, dp[i+1][column-1] - dungeon[i][column-1]);
            for(int j = column-2;j>=0;j--){
                dp[i][j] = std::max(1, std::min(dp[i+1][j], dp[i][j+1]) - dungeon[i][j]);
            }
        }
        return dp[0][0];
    }
};

int main(){
    int test[][3] = {{-2, -3, 3}, {-5, -10, 1}, {10, 30, -5}};
    std::vector<std::vector<int> > dungeon;
    for (int i = 0; i < 3; i++){
        dungeon.push_back(std::vector<int>());
        for (int j = 0; j < 3; j++){
            dungeon[i].push_back(test[i][j]);
        }
    }
    Solution solve;
    printf("%d\n", solve.calculateMinimumHP(dungeon));
    return 0;
}

File: solution/174/solution.py
class Solution(object):
    def calculateMinimumHP(self, dungeon):
        """
        :type dungeon: List[List[int]]
        :rtype: int
        """
        if len(dungeon) == 0:return 0
        row = len(dungeon)
        column = len(dungeon[0])
        result = [[0]*column for _ in xrange(row)]
        result[-1][-1] = max(1, 1-dungeon[-1][-1])
        for i in xrange(column-2, -1, -1):
            result[-1][i] = max(1, result[-1][i+1] - dungeon[-1][i])
        for i in xrange(row-2, -1, -1):
            result[i][-1] = max(1, result[i+1][-1] - dungeon[i][-1])
            for j in xrange(column-2, -1, -1):
                result[i][j] = max(1, min(result[i+1][j], result[i][j+1]) - dungeon[i][j])
        return result[0][0]